File: src/utils/url-safety.test.ts
import { describe, expect, it } from 'vitest';
import {
  UnsafeUrlError,
  checkFetchedUrl,
  isPubliclyRoutableHost,
  safeImportUrl,
} from './url-safety';

describe('what counts as a public host', () => {
  it('allows an ordinary domain', () => {
    for (const host of ['example.com', 'www.zara.com', 'shop.example.co.uk', 'xn--80ak6aa92e.com']) {
      expect(isPubliclyRoutableHost(host)).toBe(true);
    }
  });

  it('allows a public IP address', () => {
    for (const host of ['8.8.8.8', '1.1.1.1', '93.184.216.34']) {
      expect(isPubliclyRoutableHost(host)).toBe(true);
    }
  });

  it('refuses this device', () => {
    for (const host of ['localhost', 'LOCALHOST', '127.0.0.1', '127.1.2.3', '[::1]', '[::]']) {
      expect(isPubliclyRoutableHost(host)).toBe(false);
    }
  });

  it('refuses every private IPv4 range', () => {
    // A phone is inside the network these name, so a page that could not reach
    // them itself would be using the app to do it.
    for (const host of [
      '10.0.0.1',        // private
      '10.255.255.255',
      '172.16.0.1',      // private
      '172.31.255.254',
      '192.168.0.1',     // the router, most likely
      '192.168.1.254',
      '169.254.169.254', // link-local, and the cloud metadata address
      '100.64.0.1',      // carrier NAT
      '0.0.0.0',         // "this network"
      '224.0.0.1',       // multicast
      '255.255.255.255',
      '198.18.0.1',      // benchmarking
      '192.0.0.1',       // protocol assignments
    ]) {
      expect(isPubliclyRoutableHost(host)).toBe(false);
    }
  });

  it('allows the public addresses that sit next to those ranges', () => {
    // The boundaries, so the ranges are not quietly wider than intended.
    for (const host of ['11.0.0.1', '172.15.255.255', '172.32.0.1', '192.167.255.255',
                        '192.169.0.1', '100.63.255.255', '100.128.0.1', '9.255.255.255']) {
      expect(isPubliclyRoutableHost(host)).toBe(true);
    }
  });

  it('refuses four numbers that are not an address', () => {
    // 8.8.8.999 is not an IPv4 address, so a resolver treats it as a hostname
    // and it means nothing. Passing it through as "public" would be reading it
    // as an address that happens not to be private.
    for (const host of ['8.8.8.999', '1.2.3.256', '999.999.999.999']) {
      expect(isPubliclyRoutableHost(host)).toBe(false);
    }
  });

  it('allows a domain spelled entirely out of hex digits', () => {
    // The catch-all for oddly written addresses has to key on the shape of the
    // labels, not the characters: these are real domains.
    for (const host of ['face.be', 'abc.de', 'cafe.fr', 'dad.ad', 'be.ee']) {
      expect(isPubliclyRoutableHost(host)).toBe(true);
    }
  });

  it('refuses an address written to look like something else', () => {
    // A resolver accepts all of these as 127.0.0.1; a check that only understood
    // dotted decimal would let them through.
    for (const host of ['0x7f.0.0.1', '2130706433', '017700000001', '127.1', '0177.0.0.1',
                        '1.2.3.4.5', '0x7f.0x0.0x0.0x1', '10.0177.0.1']) {
      expect(isPubliclyRoutableHost(host)).toBe(false);
    }
  });

  it('refuses an IPv4 address dressed as IPv6', () => {
    expect(isPubliclyRoutableHost('[::ffff:127.0.0.1]')).toBe(false);
    expect(isPubliclyRoutableHost('[::ffff:192.168.1.1]')).toBe(false);
  });

  it('allows a public IPv6 address', () => {
    expect(isPubliclyRoutableHost('[2606:2800:220:1:248:1893:25c8:1946]')).toBe(true);
  });

  it('refuses private and link-local IPv6', () => {
    for (const host of ['[fc00::1]', '[fd12:3456::1]', '[fe80::1]', '[fe80::abcd]', '[feb0::1]']) {
      expect(isPubliclyRoutableHost(host)).toBe(false);
    }
  });

  it('refuses the names a local network gives itself', () => {
    for (const host of ['printer.local', 'nas.lan', 'api.internal', 'router.home.arpa', 'app.localhost']) {
      expect(isPubliclyRoutableHost(host)).toBe(false);
    }
  });

  it('refuses a bare intranet name', () => {
    // "router" or "nas" resolves to whatever the phone's network says it is. A
    // real product page has a domain.
    for (const host of ['router', 'nas', 'intranet', 'wiki']) {
      expect(isPubliclyRoutableHost(host)).toBe(false);
    }
  });

  it('ignores a trailing dot and surrounding space', () => {
    expect(isPubliclyRoutableHost('example.com.')).toBe(true);
    expect(isPubliclyRoutableHost(' 127.0.0.1 ')).toBe(false);
    expect(isPubliclyRoutableHost('')).toBe(false);
  });
});

describe('normalizing a URL for import', () => {
  it('accepts a product page and fills in the scheme', () => {
    expect(safeImportUrl('zara.com/uk/shirt-p123.html'))
      .toBe('https://zara.com/uk/shirt-p123.html');
  });

  it('keeps the query, drops the fragment', () => {
    expect(safeImportUrl('https://example.com/p?id=7#reviews'))
      .toBe('https://example.com/p?id=7');
  });

  it('refuses a scheme that is not a web page', () => {
    for (const url of ['file:///etc/passwd', 'ftp://example.com/x', 'javascript:alert(1)',
                       'content://media/external/images/1', 'data:text/html,hi']) {
      expect(() => safeImportUrl(url)).toThrow(UnsafeUrlError);
    }
  });

  it('refuses a URL naming this device or the local network', () => {
    for (const url of ['http://localhost:8080/admin', 'http://192.168.1.1/',
                       'http://169.254.169.254/latest/meta-data/', 'http://[::1]:3000/']) {
      expect(() => safeImportUrl(url)).toThrow(UnsafeUrlError);
    }
  });

  it('names the host it refused, so the message is worth reading', () => {
    expect(() => safeImportUrl('http://192.168.1.1/')).toThrow(/192\.168\.1\.1/);
  });

  it('refuses a URL carrying credentials', () => {
    // https://real.example@evil.test reads as the first host and fetches the
    // second, which is the whole point of writing it that way.
    expect(() => safeImportUrl('https://zara.com@evil.test/p')).toThrow(UnsafeUrlError);
    expect(() => safeImportUrl('https://user:pass@example.com/p')).toThrow(UnsafeUrlError);
  });

  it('refuses nothing at all', () => {
    expect(() => safeImportUrl('')).toThrow(UnsafeUrlError);
    expect(() => safeImportUrl('   ')).toThrow(UnsafeUrlError);
    expect(() => safeImportUrl('not a url')).toThrow(UnsafeUrlError);
  });
});

describe('checking where a request ended up', () => {
  it('accepts a redirect to another public page', () => {
    expect(() =>
      checkFetchedUrl('https://www.example.com/p', 'https://example.com/p')
    ).not.toThrow();
  });

  it('refuses a redirect onto the local network', () => {
    // The reason this exists: the initial URL passing says nothing about where
    // it sends you. By the time this runs the request has been made, so what it
    // stops is the response being read -- nothing comes back out of it.
    expect(() => checkFetchedUrl('http://192.168.1.1/admin', 'https://example.com/p'))
      .toThrow(UnsafeUrlError);
    expect(() => checkFetchedUrl('http://localhost:9000/', 'https://example.com/p'))
      .toThrow(UnsafeUrlError);
  });

  it('refuses a redirect to another scheme', () => {
    expect(() => checkFetchedUrl('file:///etc/hosts', 'https://example.com/p'))
      .toThrow(UnsafeUrlError);
  });

  it('says nothing when the platform did not report a final URL', () => {
    // Some platforms leave response.url empty. The requested URL was already
    // checked, so there is nothing further to say.
    expect(() => checkFetchedUrl(null, 'https://example.com/p')).not.toThrow();
    expect(() => checkFetchedUrl('', 'https://example.com/p')).not.toThrow();
    expect(() => checkFetchedUrl(undefined, 'https://example.com/p')).not.toThrow();
  });
});

File: domain/src/test/kotlin/com/wardrobapp/domain/UrlSafetyTest.kt
package com.wardrobapp.domain

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * Which addresses the app will fetch, and what it says about the rest.
 *
 * The check exists because a link arrives from outside -- a `wardrobapp://` deep
 * link or the share sheet -- and the phone running the app sits *inside* a home
 * network. Without this, a web page could use the app to reach a router, a
 * printer or a NAS that the page cannot reach itself.
 *
 * So the cases below are written as a table of edges rather than a handful of
 * examples: every private range appears with the address on each side of its
 * boundary, because a range written one octet out is the kind of mistake that
 * reads correctly and fails silently. They replace a 180-case corpus recorded
 * from the TypeScript this was ported from; the corpus is gone with that app, and
 * these are the cases it was covering.
 */
class UrlSafetyTest {

    @Test
    fun `a public name or address is routable`() {
        for (host in listOf(
            "example.com",
            "shop.example.co.uk",
            "8.8.8.8",
            "1.1.1.1",
            "9.255.255.255",
            "11.0.0.0",
            "172.15.255.255",
            "172.32.0.0",
            "192.167.255.255",
            "192.169.0.0",
            "126.255.255.255",
            "128.0.0.0",
            "[2001:db8::1]",
            // A domain spelled entirely out of hex characters is still a domain.
            "face.be",
        )) {
            assertTrue(isPubliclyRoutableHost(host), "$host should be routable")
        }
    }

    @Test
    fun `this device and its network are not`() {
        for (host in listOf(
            // Loopback, and the whole of 127/8 rather than one address in it.
            "127.0.0.1",
            "127.255.255.254",
            "0.0.0.0",
            // The three private ranges, first and last address of each.
            "10.0.0.0",
            "10.255.255.255",
            "172.16.0.0",
            "172.31.255.255",
            "192.168.0.0",
            "192.168.255.255",
            // Link-local, including the cloud metadata address.
            "169.254.0.1",
            "169.254.169.254",
            // Carrier-grade NAT.
            "100.64.0.0",
            "100.127.255.255",
            // IPv6 loopback, link-local and unique-local.
            "[::1]",
            "[fe80::1]",
            "[fc00::1]",
            "[fd12:3456::1]",
            // Names a resolver answers from the local network.
            "localhost",
            "router",
            "nas.local",
            "server.lan",
            "printer.internal",
            // Addresses written the ways a resolver still accepts.
            "0x7f.0.0.1",
            "0177.0.0.1",
            "127.1",
            "2130706433",
        )) {
            assertFalse(isPubliclyRoutableHost(host), "$host should be refused")
        }
    }

    @Test
    fun `an empty or trailing-dot host is handled rather than crashing`() {
        assertFalse(isPubliclyRoutableHost(""))
        assertFalse(isPubliclyRoutableHost("   "))
        assertTrue(isPubliclyRoutableHost("example.com."))
        // Case is not a way past the check.
        assertFalse(isPubliclyRoutableHost("LocalHost"))
        assertFalse(isPubliclyRoutableHost("NAS.LOCAL"))
    }

    @Test
    fun `an accepted address comes back normalized`() {
        assertEquals("https://example.com/", safeImportUrl("https://example.com"))
        assertEquals("https://example.com/p", safeImportUrl("  https://example.com/p  "))
        // A bare host is assumed to be https rather than refused.
        assertEquals("https://example.com/", safeImportUrl("example.com"))
        // The default port goes; a non-default one stays.
        assertEquals("https://example.com/", safeImportUrl("https://example.com:443"))
        assertEquals("https://example.com:8443/", safeImportUrl("https://example.com:8443"))
        // The fragment is dropped -- it is never sent to a server anyway -- and
        // the query is kept, because that is where a product id lives.
        assertEquals("https://example.com/p?id=7", safeImportUrl("https://example.com/p?id=7#top"))
        // Dot segments are resolved rather than passed on.
        assertEquals("https://example.com/b", safeImportUrl("https://example.com/a/../b"))
    }

    @Test
    fun `each refusal says which kind of refusal it is`() {
        val cases = listOf(
            "" to UnsafeUrlReason.UrlRequired,
            "   " to UnsafeUrlReason.UrlRequired,
            "https://" to UnsafeUrlReason.NotAWebAddress,
            "ftp://example.com/x" to UnsafeUrlReason.SchemeNotAllowed,
            "wardrobapp://import" to UnsafeUrlReason.SchemeNotAllowed,
            // Not even an address: there is no `//`, so parsing fails before the
            // scheme is judged. Either refusal is correct; this pins which one.
            "javascript:alert(1)" to UnsafeUrlReason.NotAWebAddress,
            // Reads as the first host, fetches the second.
            "https://real.example@evil.test/" to UnsafeUrlReason.CredentialsInUrl,
            "https://user:pw@example.com/" to UnsafeUrlReason.CredentialsInUrl,
            "http://192.168.1.1/" to UnsafeUrlReason.HostIsLocal("192.168.1.1"),
            "https://localhost:3000/" to UnsafeUrlReason.HostIsLocal("localhost"),
        )

        for ((input, expected) in cases) {
            val thrown = assertFailsWith<UnsafeUrlException>("$input was not refused") {
                safeImportUrl(input)
            }
            assertEquals(expected, thrown.reason, "for $input")
        }
    }

    @Test
    fun `a refusal carries the host it refused`() {
        // The host is what makes the message worth reading: "that link points
        // somewhere I will not go" is not actionable, and the name is.
        val thrown = assertFailsWith<UnsafeUrlException> { safeImportUrl("http://10.0.0.5/cam") }

        assertEquals(UnsafeUrlReason.HostIsLocal("10.0.0.5"), thrown.reason)
        assertTrue("10.0.0.5" in thrown.reason.englishMessage())
    }

    @Test
    fun `a redirect onto the local network is refused after the fact`() {
        // The second line of defence. On Android the request is stopped before it
        // is made, but the response is checked too: this is what stops one being
        // read and parsed.
        val thrown = assertFailsWith<UnsafeUrlException> {
            checkFetchedUrl("http://192.168.1.1/admin", "https://example.com/p")
        }

        assertEquals(UnsafeUrlReason.RedirectedToLocalHost("192.168.1.1"), thrown.reason)
    }

    @Test
    fun `a redirect that cannot be read is refused too`() {
        val thrown = assertFailsWith<UnsafeUrlException> {
            checkFetchedUrl("not a url at all", "https://example.com/p")
        }

        assertEquals(UnsafeUrlReason.RedirectUnreadable, thrown.reason)
    }

    @Test
    fun `a redirect within the public web is allowed`() {
        // No exception is the assertion.
        checkFetchedUrl("https://www.example.com/p", "https://example.com/p")
        checkFetchedUrl("https://example.com/p", "https://example.com/p")
        // Nothing to check: some platforms report no final URL at all, and the
        // requested one was checked before the request went out.
        checkFetchedUrl(null, "https://example.com/p")
        checkFetchedUrl("", "https://example.com/p")
    }

    @Test
    fun `a redirect off the web is refused, with the local-network wording`() {
        // Worth pinning because it looks like a bug and is not: a public host on
        // a scheme that is not the web shares the redirect refusal, so the
        // sentence mentions the network. The outcome -- refusal -- is right, and
        // the alternative is a reason that exists for one unreachable case.
        val thrown = assertFailsWith<UnsafeUrlException> {
            checkFetchedUrl("ftp://example.com/x", "https://example.com/p")
        }

        assertEquals(UnsafeUrlReason.RedirectedToLocalHost("example.com"), thrown.reason)
    }

    @Test
    fun `every reason has a sentence, and no two share one`() {
        // The reasons are values so :app can translate them; the English is kept
        // for the places that have no resources. A reason with no sentence, or two
        // reasons with the same one, would make a refusal unreadable.
        val reasons = listOf(
            UnsafeUrlReason.UrlRequired,
            UnsafeUrlReason.NotAWebAddress,
            UnsafeUrlReason.SchemeNotAllowed,
            UnsafeUrlReason.CredentialsInUrl,
            UnsafeUrlReason.HostIsLocal("192.168.1.1"),
            UnsafeUrlReason.RedirectUnreadable,
            UnsafeUrlReason.RedirectedToLocalHost("192.168.1.1"),
        )

        val messages = reasons.map { it.englishMessage() }

        assertTrue(messages.all { it.isNotBlank() }, "a reason has no sentence")
        assertEquals(messages.size, messages.toSet().size, "two reasons read the same")
    }
}
